File: 2025/42. Trapping Rain Water/Solution.java
class Solution {
    public int trap(int[] height) {
        int left = 0, right = height.length - 1;
        int maxLeft = height[left], maxRight = height[right];
        int sum = 0;
        while (left <= right) {
            if (maxLeft <= maxRight) {
                if (height[left] < maxLeft) {
                    sum += maxLeft - height[left];
                } else {
                    maxLeft = height[left];
                }
                left ++;
            } else {
                if (height[right] < maxRight) {
                    sum += maxRight - height[right];
                } else {
                    maxRight = height[right];
                }
                right --;
            }
        }
        return sum;
    }
}

File: 2025/42. Trapping Rain Water/solution.py
class Solution:
    def trap(self, height: List[int]) -> int:
        lm, rm = height[0], height[-1]
        l, r, max_water = 0, len(height) - 1, 0
        while l <= r:
            if lm < rm:
                lm = max(lm, height[l])
                max_water += lm - height[l]
                l += 1
            else:
                rm = max(rm, height[r])
                max_water += rm - height[r]
                r -= 1
        return max_water